### ISSUE TITLE
lambda delete action - handle lambda @ edge functions

### ISSUE THREAD

**Issue by kapilt** [capitalone/cloud-custodian#1688]
`
An error occurred (InvalidParameterValueException) when calling the DeleteFunction operation: Lambda was unable to delete arn:aws:lambda:us-east-1:xxx:function:xxx-Test-LambdaEdge:1 because it is a replicated function.: InvalidParameterValueException`

**Comment by lfreneda** [comment]
+1

**Comment by zacharya** [comment]
Looks like AWS is handling the deletion of replicated functions behind the scenes: 
 https://forums.aws.amazon.com/ann.jspa?annID=5529

The way you trigger this is by removing the CloudFront associations which then kicks off the deletion of the replicated functions automatically, but until that completes you can't delete the parent function.

AWS says the replicated functions are "typically deleted within a few hours".  Does c7n have a framework we could plug into for handling deferred actions that periodically check and wait on a precondition before processing?

**Comment by thisisshi** [comment]
@zacharya the best way to do it would be to use the `mark-for-op` pattern I suppose, pseudocode policy:

```yaml
policies:
  - name: mark-lambda-edge-for-deletion
    resource: lambda
    actions:
      - delete-associated-cloudfront-distributions # not a real action, just spitballling
      - type: mark-for-op
        tag: c7n_lambda_edge_delete
        op: delete
        days: 1

  - name: delete-lambda-edge
    resource: lambda
    filters:
      - type: marked-for-op
        tag: c7n_lambda_edge_delete
      - no-cloudfront-distributions # not a real filter either
    actions:
      - delete
```

**Comment by zacharya** [comment]
Makes sense.

The only possible issue with this approach is that you'd have to crawl all CloudFront distros until you find any associated with the Lambda function due to how the CloudFront API is designed (https://boto3.readthedocs.io/en/latest/reference/services/cloudfront.html#CloudFront.Client.list_distributions).  There's no way to filter on Lambda associations nor is there an endpoint in the Lambda API that identifies which CloudFront distributions are associated with the function.

I'm happy to take this as a first contribution and submit a PR if no one objects.

**Comment by thisisshi** [comment]
Yeah, looks like it would we would need the following:
- filter: `has-cloudfront-distribution`
  - list-cloudfront-distributions
  - get-cloudfront-distributions
    - look at the lambdafunctionassociations object
- action: `delete-cloudfront-distribution`
  - this would have to do the same thing as the filter unfortunately from what I can see
